### ISSUE TITLE
1.11.x readiness probe change causes prometheus alerts

### ISSUE THREAD

**Issue by taxilian** [rook/rook#11791]
**Is this a bug report or feature request?**
* More of an issue to be discussed =]

**Deviation from expected behavior:**
The backup manager pod never shows "ready", so any monitoring system which is watching to alert of pods which don't come up correctly will throw alerts. Additionally it always shows the state as "1/2" instead of "2/2" -- I doubt I'm the only one accustomed to seeing that and thinking that it means something is broken.

**Expected behavior:**
Generally people expect that if things are working correctly the "readiness" state will at 100%. The biggest issue to me is that there isn't a way I can think of to monitor that those pods are working correctly without having this alert come up.

**How to reproduce it (minimal and precise):**

* Install 1.11.0
* Install the helm chart `prometheus-community/kube-prometheus-stack` from https://prometheus-community.github.io/helm-charts with alertmanager enabled
* Wait 15 minutes and check alertmanager to see the alert. 

![image](https://user-images.githubusercontent.com/224334/221927328-163e2a65-e3b9-407b-a527-41b9250d47ba.png)

-----

I'm honestly not sure if this needs to be changed or not, but it definitely should be called out very clearly. I read the release notes and saw that you made the change but didn't realize the implications until my alerts started going off and I asked about it on the discord server.  I totally understand why the change was made and I like the idea, but I expect it to generate a lot of confusion.

**Comment by travisn** [comment]
@rkachach Could you take a look at what we can do about these alerts? Is there any way we can tell prometheus this is an expected state? It's best if we can find a fix without suppressing the alert, but we at least need to find a fix so everyone with prometheus configured doesn't hit this issue by default.

**Comment by taxilian** [comment]
The Alert rule definition is documented here: https://runbooks.prometheus-operator.dev/runbooks/kubernetes/kubedeploymentreplicasmismatch/

The actual rule is:

```
alert: KubeDeploymentReplicasMismatch
annotations:
  description: Deployment {{ $labels.namespace }}/{{ $labels.deployment }}
    has not matched the expected number of replicas for longer than 15 minutes.
  runbook_url: https://runbooks.prometheus-operator.dev/runbooks/kubernetes/kubedeploymentreplicasmismatch
  summary: Deployment has not matched the expected number of replicas.
expr: |-
  (
    kube_deployment_spec_replicas{job="kube-state-metrics", namespace=~".*"}
      >
    kube_deployment_status_replicas_available{job="kube-state-metrics", namespace=~".*"}
  ) and (
    changes(kube_deployment_status_replicas_updated{job="kube-state-metrics", namespace=~".*"}[10m])
      ==
    0
  )
for: 15m
labels:
  severity: warning
```

Looks like it's using `[kube-state-metrics](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/deployment-metrics.md)` to compare if `kube_deployment_spec_replicas > kube_deployment_status_replicas_available` for more than 15 minutes.

Based on the [release notes](https://kubernetes.io/blog/2021/04/13/kube-state-metrics-v-2-0/#why), this appears to be the recommended way to check "whether a deployment is rolled out successfully or stuck." for alerting.

**Comment by taxilian** [comment]
Hopefully one of you folks can see something I missed, but I'm not seeing any way to prevent this from firing -- all I can see to do is to silence that particular warning for the mgr pods, which means there will be no warnings if there actually is a problem with starting them. :-(

**Comment by travisn** [comment]
The ceph standby mgr actually does forward requests to the active mgr by default. This just doesn't work when running in a rook/k8s environment because the expected forwarding urls don't work. 
@rkachach Could you investigate if there is some change we could make so the requests could be forwarded properly in our environment? This would certainly be ideal, then the readiness probe wouldn't fail and cause these side effects. 

If this takes some time to work out, we may need to consider reverting to use the sidecar again for now.

**Comment by rkachach** [comment]
@travisn I'm trying to see if there's some solution to the alarm issue. I don't think we can acheive the same behaviour without using the `readiness probe`.

Anyway, just to give some more context, to clarify the reason behind the change and to answer some concerns:

1. The issue we are trying to solve here is a long standing problem and a known k8s issue [2]. Readiness probe is among the prposed solutions that some people used to address the problem.

1. The previous HA implementation (based on playing with labels to indicate which mgr is the active one) suffered from several issues and BUGs such as: https://github.com/rook/rook/issues/11596 https://github.com/rook/rook/issues/11759 https://github.com/rook/rook/issues/11465 https://github.com/rook/rook/issues/11147 https://github.com/rook/rook/issues/10914 https://github.com/rook/rook/issues/10796  https://github.com/rook/rook/issues/9182 .... and other similar problems. So if I have to choose between a buggy solution that doesn't work in several situations and another that works but raises a `false alarm` IMHO as a user I would prefer the second. Obviously the ideal solution would be something that works correctly without leading to any false alarm. Anyway this is something that we can discuss to see what's the best option. 

1. Ceph `mgr` high-availability mechanism is based on `1 active - N standby` algorithm. At any moment only the active can process the traffic, the standby are there ready to perform a quick fail-over if necessary but they are not meant to process any kind of traffic. Active/Standby elecction is an internal mechanism managed by ceph and not k8s. As travisn explained, ceph `mgr standby` already tries to `redirect` the trafic to the active one. This redirection implies generating a url pointing to the active manager pod. This works on other environments but not in k8s as unfortunately right now this URL points to the pods `ClusterIP` which is internal. If you have a Node Port service (providing a ceph-dashboard i.e) then k8s could redirect the request to all the ready pods including the `standby mgr`, but since this is redirecting to a cluster IP the browser will hangup as it cannot reach this internal URL from the outside. 

5. The k8s definition of readiness probe [1] states: `The kubelet uses readiness probes to know when a container is ready to start accepting traffic.` . In this case the ceph standby manager was not designed to process traffic. I mean technically it's not ready. Worse, the redirection mechanism doesn't work in k8s environment as I explained in the point (1. So in fact it's not a misuse of the readiness probe, it just reflects the reality. 

6. For k8s when a pod is ready, it means that it can process traffic [1]. So for k8s it doesn't mean that `i am ready to do my job`. It means `I can process traffic` and by consequence k8s service will load-balance the requests between all the available pods regardless of wether they are active or standby. It would have been great to solve our issue if k8s actualy differentiate between `Im ready/OK` and `I can process traffic`. Unfortunatelly this is not the case and the readiness combines both conditions. I'm not aware of a solution to this specific problem, if somebody knows some other solution for this case I'm open to investigate it.

In summary, I'll recommend to adapt the prometheus alarm to take into account the ceph mgr behaviour so to do not raise an alarm in this specific case. Going back to the previous HA implementation has the risk of running into all the BUG (and other similar ones) described in the point 1). 

[1] https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
[2] https://github.com/kubernetes/kubernetes/issues/45300

**Comment by rkachach** [comment]
@taxilian 

I'm not sure if modifying the rules is an option in your case. If it's, just to let you know that I was experementing on a small cluster with the idea of adjusting the rules so to do not fire an alarm for the ceph mgr deployments. I tested the following rules and went go on my small clusater so far. Please, can you double check if using these rules (or something similar) solves your issue? Thanks.

```
    - alert: KubeDeploymentReplicasMismatch
      annotations:
        description: Deployment {{ $labels.namespace }}/{{ $labels.deployment }} has
          not matched the expected number of replicas for longer than 15 minutes.
        runbook_url: https://runbooks.prometheus-operator.dev/runbooks/kubernetes/kubedeploymentreplicasmismatch
        summary: Deployment has not matched the expected number of replicas.
      expr: |-
        (
          kube_deployment_spec_replicas{job="kube-state-metrics", namespace=~".*", deployment!~".*rook-ceph-mgr-.*"}
            >
          kube_deployment_status_replicas_available{job="kube-state-metrics", namespace=~".*", deployment!~".*rook-ceph-mgr-.*"}
        ) and (
          changes(kube_deployment_status_replicas_updated{job="kube-state-metrics", namespace=~".*", deployment!~".*rook-ceph-mgr-.*"}[10m])
            ==
          0
        )
      for: 15m
      labels:
        severity: warning
    - alert: KubeCephMgrDeploymentReplicasMismatch
      annotations:
        description: Deployment {{ $labels.namespace }}/{{ $labels.deployment }} has
          not matched the expected number of replicas for longer than 1 minute.
        summary: Not enough ceph mgr ready deployments.
      expr: sum(kube_deployment_status_replicas_ready{deployment=~".*rook-ceph-mgr-.*"}) < 1
      for: 1m
      labels:
        severity: warning
```

**Comment by taxilian** [comment]
That rule looks like it would work, but is there a meaningful difference between doing that and just silencing that particular rule for that deployment? My main concern is that there isn't really a good way to tell if something else is wrong with the deployment -- e.g. I don't think there would be a visible difference between the pod not starting up for some other reason and not becoming fully ready.

It's something I can tolerate on my system, but it's something I wanted to surface because I expect it to come up repeatedly for as long as you're using this method.

**Comment by travisn** [comment]
@rkachach Thanks for the investigation in the different considerations. While I do agree the current implementation is superior for handling the active/standby mgr, I don't think we can keep this behavior by default. The main reasons are:
- Rook doesn't own alerts that are firing so we can't fix it automatically.
- If we left this enabled, everyone that enables monitoring will see the alert and believe that something is broken by default. Even if we document the recommended alert to update, some people will miss it and be concerned about the failure even though it is expected.
- The immediate reaction we have from users after the v1.11.0 release shows how important it is that we don't raise this alert by default.

If we went back to the sidecar implementation, there are a couple of shortcomings, but I don't think most of those issues referenced in the first point of [this comment](https://github.com/rook/rook/issues/11791#issuecomment-1451710899) are directly related. For example, the mgr sidecar was first introduced in v1.7.11, but one of the issues referenced v1.7.7, and some of the other issues were closed by the stalebot so they didn't have root cause analysis.

I'm thinking we can fallback to a hybrid approach:
- Add back the sidecar implementation and enable it by default
- Add a setting to the cephcluster CR to allow enabling of the readiness probe (which would disable the sidecar)
- In Rook's monitoring documentation, recommend that they enable the readiness probe and disable the related alert. This way, only people who pay attention to the documentation will ever see the alert and will immediately know it's expected.

Even in [1] there are designs proposed that involve operator solutions, instead of relying solely on the readiness probe. But it definitely seems like an interesting pattern to follow up with on the community.

[1] https://github.com/kubernetes/kubernetes/issues/45300#issuecomment-737658596

**Comment by reefland** [comment]
It's more than just Prometheus. This is also now a never ending deployment within ArgoCD, it never reaches completion.  It's just stuck in limbo:
![image](https://user-images.githubusercontent.com/16230914/223242704-37ba7a8a-b250-49f2-adf9-a9c75eefb100.png)

And this triggers other ArgoCD alerts about deployment being stuck.  And wasting resources waiting for it....
![image](https://user-images.githubusercontent.com/16230914/223246716-02bbbf23-57a4-4f07-b175-f09aa561cda4.png)

**Comment by rkachach** [comment]
By now, taking into considerations the issues caused by the readiness probe I'll just revert the related changes in the next release. Later on we will try to fix the related HA issues by using another method.
